db/snapshot/backup_task.cc

```diff
@@ -103,52 +105,115 @@ future<> backup_task_impl::upload_component(shared_ptr<s3::client>& client, sstr
             }
-        }
-        if (!component_ent.has_value()) {
+        } catch (...) {
+            ex = std::current_exception();
             break;
         }
-        auto gh = uploads.hold();
+    } while (component_ent);
+    co_await snapshot_dir_lister.close();
+    if (ex)
+        co_await coroutine::return_exception_ptr(std::move(ex));
 
-        // Pre-upload break point. For testing abort in actual s3 client usage.
-        co_await utils::get_local_injector().inject("backup_task_pre_upload", utils::wait_for_message(std::chrono::minutes(2)));
+    std::ranges::sort(backup_files, std::greater());
+    co_return backup_files;
+}
+} // namespace
 
-        std::ignore = upload_component(cln, component_ent->name).handle_exception([&ex] (std::exception_ptr e) {
-            // keep the first exception
-            if (!ex) {
-                ex = std::move(e);
-            }
-        }).finally([gh = std::move(gh)] {});
-        co_await coroutine::maybe_yield();
-        co_await utils::get_local_injector().inject("backup_task_pause", utils::wait_for_message(std::chrono::minutes(2)));
-        if (impl::_as.abort_requested()) {
-            ex = impl::_as.abort_requested_exception_ptr();
-            break;
-        }
+future<> backup_task_impl::do_backup() {
+    if (!co_await file_exists(_snapshot_dir.native())) {
+        throw std::invalid_argument(fmt::format("snapshot does not exist at {}", _snapshot_dir.native()));
     }
 
-    co_await snapshot_dir_lister.close();
-    co_await uploads.close();
-    if (ex) {
-        co_await coroutine::return_exception_ptr(std::move(ex));
+    auto backup_files = co_await get_backup_files(_snapshot_dir);
+    std::atomic_size_t counter = 0;
+
+    sharded<abort_source> sharded_aborter;
+    co_await sharded_aborter.start();
+    gate as_gate;
+    auto s = std::make_unique<optimized_optional<abort_source::subscription>>(_as.subscribe([&]() noexcept {
+        auto h = as_gate.hold();
+        std::ignore = sharded_aborter.invoke_on_all([ex = _as.abort_requested_exception_ptr()](abort_source& aborter) {
+            aborter.request_abort_ex(ex);
+        }).finally([h = std::move(h)] {});
+    }));
+
+    co_await sharded_aborter
+        .invoke_on_all([this, &backup_files, &counter, &sharded_aborter](abort_source& aborter) -> future<> {
+            auto shard_id = this_shard_id();
+            gate uploads;
+            std::exception_ptr ex;
+            try {
+                auto cln = _snap_ctl.storage_manager().container().local().get_endpoint_client(_endpoint);
+
+                // Fetching files sequentially within each shard allows less loaded shards to take on more work, compensating for those that are lagging behind.
+                auto backup_file_idx = [&counter, backup_size = backup_files.size()] -> std::optional<size_t> {
+                    auto ret_val = counter++;
+                    if (ret_val < backup_size) {
```

Comment:
It's not clear why you convert `counter++ < backup_files.size()` condition into `std::optional<size>::has_value()` one. I mean, lines 165-173 (i.e. the backup_file_idx lambda and while() loop below) could look as simple as

```c++
    while (auto name_idx = counter.fetch_add(1); name_idx < backup_files.size()) {
```